File: 2658-maximum-number-of-fish-in-a-grid/2658-maximum-number-of-fish-in-a-grid.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var findMaxFish = function (grid) {
    const dx = [0, 0, 1, -1];
    const dy = [1, -1, 0, 0];
    const n = grid.length, m = grid[0].length;
    const isValid = (i, j) => i >= 0 && i < n && j >= 0 && j < m;
    const dfs = (i, j) => {
        if (!isValid(i, j) || !grid[i][j]) return 0;
        let cnt = grid[i][j];
        grid[i][j] = 0;
        for (let k = 0; k < dx.length; ++k) {
            const iNew = dx[k] + i;
            const jNew = dy[k] + j;
            cnt += dfs(iNew, jNew);
        }
        return cnt;
    }
    let ans = 0;
    for (let i = 0; i < n; ++i) {
        for (let j = 0; j < m; ++j) ans = Math.max(ans, dfs(i, j));
    }
    return ans;
};

File: 2658-maximum-number-of-fish-in-a-grid/2658-maximum-number-of-fish-in-a-grid.cs
public class Solution
{
    int[] dx = new int[] { 0, 0, 1, -1 };
    int[] dy = new int[] { 1, -1, 0, 0 };
    int n, m;
    bool isValid(int i, int j) { return i >= 0 && i < n && j >= 0 && j < m; }
    int dfs(int i, int j, int[][] grid)
    {
        if (!isValid(i, j) || grid[i][j] == 0) return 0;
        int cnt = grid[i][j];
        grid[i][j] = 0;
        for (int k = 0; k < dx.Length; ++k)
        {
            int iNew = dx[k] + i;
            int jNew = dy[k] + j;
            cnt += dfs(iNew, jNew, grid);
        }
        return cnt;
    }
    public int FindMaxFish(int[][] grid)
    {
        n = grid.Length;
        m = grid[0].Length;

        int ans = 0;
        for (int i = 0; i < n; ++i)
        {
            for (int j = 0; j < m; ++j) ans = Math.Max(ans, dfs(i, j, grid));
        }
        return ans;
    }
}